Real magic button

andrewn opened the pull request:
Enabling the physical magic button.

Splits out the actions into two types. `atomic` actions which are short-lived and `magic` actions which have a start and stop state.

_Thread for discussion_

andrewn committed:
Device instance has handle method that proxies to state machine

andrewn committed:
Non-namespaced 'atomic' action parsing

andrewn committed:
Validate action type for magic functions that have start/stop events

andrewn committed:
Device provides settings to action handler when it's run

andrewn committed:
Upgrade machina module

andrewn committed:
Set currentMagicAction for status API

andrewn committed:
Remove console.log

andrewn committed:
Emit announcer events over event stream

andrewn committed:
Provide requested settings to all action types

andrewn committed:
Validate 'magic' action types and assume others are atomic

andrewn committed:
'magic' state action that triggers the user-configured magic action

Actions are injected with the device instance on creation allowing
them to trigger handle(), which is magic-action aware. The default
magic-action is set as 'announcer'.

andrewn committed:
Wrap state machine's transition function

pixelblend commented:
Initial thoughts, as discussed:

There are currently 2 types of actions, either of which could be assigned as functionality for the magic button:

## Short-lived actions that are too small to be stopped
Examples:
* Tweeting
* Favouriting
* Skipping a track

Possible names for this action type:
* `atomic`
* `small`
* `datagram`
* `foreground`

## Actions that have a longer life-span and can be cancelled if required.
Examples:
* Avoiding
* Announcing

Possible names for this action type:
* `stagable`
* `cancellable`
* `daemonised`

** Note that your radio could be both avoiding and announcing, although we've previously assigned these as mutually exclusive magic button functions **

andrewn committed:
Reinstate previous magic button colour when announcer stops

andrewn commented:
Also consider `toggleable` / `discrete` ?

andrewn committed:
Use stable rabbit version of client library

andrewn committed:
Note about setting development env var

andrewn committed:
Show announcer button in web UI

- get announcer state after other state loaded
- fix bug setting announcer state in data store

andrewn committed:
Implement Avoider using new start/stop state types

- /avoider root returns state info
- tracks are avoided using nowPlaying metadata
- programmes are avoided using nowAndNext metadata
- log params action invoked when debug flag is set
- expose avoid button in Web UI

andrewn committed:
Announcer route index returns state

andrewn committed:
Ensure log level set as soon as it's known

andrewn committed:
Proxy current state from State to Device instance

andrewn committed:
Cancel avoiding when switching stations

andrewn committed:
Cancel avoid when entering standby

andrewn committed:
Remove old avoid route code
